README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggsunburst

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of ggsunburst is to create a sunburst plot without much effort
while allowing to create more complex plots with a bit of work.

## Installation

You can install the development version of ggsunburst from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("sachijay/ggsunburst")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(ggsunburst)

in_dat <- tibble::tibble(
  level1_var = c(
    rep("l1 A", times = 3),
    rep("l1 B", times = 4)
  ),
  level2_var = c(
    "l2 A", "l2 A", "l2 B",
    "l2 C", "l2 C", "l2 D", "l2 D"
  ),
  level3_var = c(
    "l3 A", "l3 B", "l3 C",
    "l3 D", "l3 E", "l3 F", "l3 G"
  ),
  values_var = c(
    123, 456, 798,
    987, 654, 321, 159
  )
)

in_dat
#> # A tibble: 7 × 4
#>   level1_var level2_var level3_var values_var
#>   <chr>      <chr>      <chr>           <dbl>
#> 1 l1 A       l2 A       l3 A              123
#> 2 l1 A       l2 A       l3 B              456
#> 3 l1 A       l2 B       l3 C              798
#> 4 l1 B       l2 C       l3 D              987
#> 5 l1 B       l2 C       l3 E              654
#> 6 l1 B       l2 D       l3 F              321
#> 7 l1 B       l2 D       l3 G              159

plot_dat <- get_sunplot_dataset(
  .dat = in_dat,
  level1_var, level2_var, level3_var,
  .count_var = values_var
)

plot_dat
#> # A tibble: 14 × 7
#>    .name .level .fill .box_min .box_max .alpha .angle
#>    <chr>  <int> <chr>    <dbl>    <dbl>  <dbl>  <dbl>
#>  1 <NA>       0 <NA>    0        1       0     -90   
#>  2 l1 A       1 l1 A    0        0.394   1      19.1 
#>  3 l1 B       1 l1 B    0.394    1       1      19.1 
#>  4 l2 A       2 l1 A    0        0.166   0.75   60.2 
#>  5 l2 B       2 l1 A    0.166    0.394   0.75  -10.7 
#>  6 l2 C       2 l1 B    0.394    0.863   0.75   43.8 
#>  7 l2 D       2 l1 B    0.863    1       0.75  -65.3 
#>  8 l3 A       3 l1 A    0        0.0352  0.667  83.7 
#>  9 l3 B       3 l1 A    0.0352   0.166   0.667  53.9 
#> 10 l3 C       3 l1 A    0.166    0.394   0.667 -10.7 
#> 11 l3 D       3 l1 B    0.394    0.676   0.667  77.5 
#> 12 l3 E       3 l1 B    0.676    0.863   0.667  -6.95
#> 13 l3 F       3 l1 B    0.863    0.955   0.667 -57.1 
#> 14 l3 G       3 l1 B    0.955    1       0.667 -81.8

draw_sunburst_plot(
  .dat = plot_dat,
  .label_txt_size = 4,
  .linewidth = 0.5,
  palette = "Dark2"
)
#> Warning: Removed 1 rows containing missing values (`geom_text()`).
```

<img src="man/figures/README-example-1.png" width="100%" />


NEWS.md
# ggsunburst 0.0.0.9000

* Completed basic functionality of the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(ggsunburst)

test_check("ggsunburst")


tests/testthat/test-calculate_alpha.R
test_that("alpha is double", {
  expect_type(calculate_alpha_wo_fill(1), "double")
})

test_that("alpha is 1 for level 1", {
  expect_equal(calculate_alpha_wo_fill(1), 1)
})

test_that("alpha is Inf for root", {
  expect_equal(calculate_alpha_wo_fill(0), Inf)
})

test_that("alpha is double", {
  expect_type(calculate_alpha(1, fill = "foo"), "double")
})

test_that("alpha is 1 for level 1", {
  expect_equal(calculate_alpha(1, fill = "foo"), 1)
})

test_that("alpha is Inf for root", {
  expect_equal(calculate_alpha(0, fill = "foo"), Inf)
})

test_that("alpha is 0 for missing fill", {
  expect_equal(calculate_alpha(1, fill = NA), 0)
})

test_that("unequal argument lengths", {
  expect_error(calculate_alpha(1:2, fill = "foo"))
})




tests/testthat/test-calculate_alpha_wo_fill.R
test_that("alpha is double", {
  expect_type(calculate_alpha_wo_fill(1), "double")
})

test_that("alpha is 1 for level 1", {
  expect_equal(calculate_alpha_wo_fill(1), 1)
})

test_that("alpha is Inf for root", {
  expect_equal(calculate_alpha_wo_fill(0), Inf)
})


R/calculate_alpha.R
#' Get the alpha level
#'
#' @param level An integer level of the sunburst plot.
#'
#' @description Maps the level of the plot to an alpha value.
#'
#' @return A number between 0.5 and 1. The maximum value is 1 (lowest/innermost level) and the minimum value is 0.5 (highest/outermost level).
calculate_alpha_wo_fill <- function(
    level
){
  
  if(!is.numeric(level)){
    
    stop("Provide an integer level!!!")
    
  }
  
  alpha <- 0.5*(1+1/level)
  
  return(alpha)
  
}


#' Get the alpha level
#' 
#' @description A wrapper for [calculate_alpha_wo_fill()] with 0 alpha values for non filled parts.
#'
#' @param level An integer level of the sunburst plot.
#' @param fill A character vector (or a factor) with the fill levels. 
#'
#' @return A number between 0.5 and 1. The maximum value is 1 (lowest/innermost level) and the minimum value is 0.5 (highest/outermost level). 0 if the fill is `NA`.
#'
#' @details The value of the fill is not important. The only importance of the argument is to check if the fill is missing or not.
calculate_alpha <- function(
    level, 
    fill
){
  
  if(length(level) != length(fill)){
    
    stop("level and fill lengths should be equal!!!")
    
  }
  
  alpha <- ifelse(
    test = is.na(fill),
    yes = 0,
    no = calculate_alpha_wo_fill(level)
  )
  
  return(alpha)
  
}


R/calculate_angle.R
#' Calculate the label angle 
#' 
#' @description Get the angle from the end points of the segment. The angles makes the segment labels perpendicular to the circle. 
#' An angle of 0 is the positive x-axis so the angle adjustment is directly the correction for ggplot. 
#' A correction is applied to make the labels on the left side of the plot more readable.
#'
#' @param box_min A numeric vector of lower endpoints of segments as a proportion.
#' @param box_max A numeric vector of upper endpoints of segments as a proportion.
#'
#' @return The angle correction as should be given to ggplot.
calculate_angle <- function(box_max, box_min){
  
  if(!is.numeric(box_max)){

    stop("'box_max' should be a numeric value or a vector.")

  }

  if(!is.numeric(box_min)){

    stop("'box_min' should be a numeric value or a vector.")

  }

  if(any(box_max < 0) | any(box_max > 1)){

    stop("'box_max' should be a proportion!!!")

  }

  if(any(box_min < 0) | any(box_min > 1)){

    stop("'box_min' should be a proportion!!!")

  }
  
  angle = 90 - (box_max + box_min)*180

  angle = ifelse(
    test = abs(angle) > 90,
    yes = 180 + angle,
    no = angle
  )
  
  return(angle)
  
}


R/get_sunplot_dataset.R
#' Get a prepared dataset to draw the sunburst plot.
#' 
#' @description Creates a dataset prepared as input to the `draw_sunburst_plot` function.
#'
#' @param .dat The data frame or a tibble with the variables and the values for the plot. See the example for the input dataset structure.
#' @param ... Variables in `.dat` that matches the order of the circles in the sunburst plot (innermost to outermost).
#' @param .count_var The variable with the values to decide the sizes of the circles. 
#' 
#' @details The returned dataset includes the following columns.  
#' - `.name`: Label of the segment.
#' - `.level`: Level of the sunburst plot. Level 1 is the innermost level.
#' - `.fill`: Value for the fill variable. This is not the final colour in the plot, but the same value will have the same colour for all levels.
#' - `.box_min` & `.box_max`: Minimum and maximum bounds for the segment.
#' - `.alpha`: The transparency for that level.
#' - `.angle`: The angle of the text label.
#'
#' @return A tibble to provide as input to the `draw_sunburst_plot` function. See details for the list of variables.
#' 
#' @export
#'
#' @examples
#' in_dat <- tibble::tibble(
#' level1_var = c(
#' rep("l1 A", times = 3),
#' rep("l1 B", times = 4)
#' ),
#' level2_var = c(
#' "l2 A", "l2 A", "l2 B", 
#' "l2 C", "l2 C", "l2 D", "l2 D"
#' ),
#' level3_var = c(
#' "l3 A", "l3 B", "l3 C", 
#' "l3 D", "l3 E", "l3 F", "l3 G"
#' ),
#' values_var = c(
#' 123, 456, 798, 
#' 987, 654, 321, 159
#' )
#' )
#' 
#' in_dat
#' 
#' 
#' get_sunplot_dataset(
#' .dat = in_dat,
#' level1_var, level2_var, level3_var,
#' .count_var = values_var
#' )
get_sunplot_dataset <- function(
    .dat,
    ...,
    .count_var
){
  
  n_levels <- rlang::dots_n(...)
  dots_vars <- rlang::enquos(...)
  
  l0 <- tibble::tibble(
    .name = NA,
    .level = 0,
    .fill = NA,
    .box_min = 0,
    .box_max = 1
  )
  
  l <- lapply(
    1:n_levels,
    function(level){
      
      dat_level_grouped <- dplyr::group_by(
        .dat,
        !!!dots_vars[1:level]
      )
      
      dat_level_summarised <- dplyr::summarise(
        .data = dat_level_grouped,
        .value = sum({{ .count_var }}),
        .groups = "drop"
      )
      
      dat_level_arranged <- dplyr::arrange(
        .data = dat_level_summarised,
        !!!dots_vars[1:level]
      )
      
      dat_level_mutated <- dplyr::mutate(
        .data = dat_level_arranged,
        .level = level,
        .fill = !!dots_vars[[1]],
        .perc = .value/sum(.value),
        .box_max = cumsum(.perc),
        .box_min = .box_max - .perc
      )
      
      dat_level <- dplyr::select(
        .data = dat_level_mutated,
        .name = !!dots_vars[[level]], 
        .level, 
        .fill, 
        .box_min, 
        .box_max
      )
      
      return(dat_level)
      
    })
  
  plot_dat_combined <- dplyr::bind_rows(l0, l)
  
  plot_dat <- dplyr::mutate(
    .data = plot_dat_combined,
    .level = as.integer(.level),
    .alpha = calculate_alpha(.level, .fill),
    .angle = calculate_angle(.box_min, .box_max)
  )
  
  return(plot_dat)
  
}


R/draw_sunburst_plot.R
#' Create the sunburst plot.
#'
#' @param .dat A data frame (or tibble) for the plot with specific requirements. Use the `get_sunplot_dataset()` to generate this dataset.
#' @param .label_txt_size A numeric value with the size of the text labels.
#' @param .linewidth Line width of the segments.
#' @param ... Passed to `ggplot2::scale_fill_brewer()` allowing to change plot colours.
#'
#' @return A ggplot object.
#' @export
#'
#' @examples
#' in_dat <- tibble::tibble(
#' level1_var = c(
#' rep("l1 A", times = 3),
#' rep("l1 B", times = 4)
#' ),
#' level2_var = c(
#' "l2 A", "l2 A", "l2 B", 
#' "l2 C", "l2 C", "l2 D", "l2 D"
#' ),
#' level3_var = c(
#' "l3 A", "l3 B", "l3 C", 
#' "l3 D", "l3 E", "l3 F", "l3 G"
#' ),
#' values_var = c(
#' 123, 456, 798, 
#' 987, 654, 321, 159
#' )
#' )
#' 
#' plot_dat <- get_sunplot_dataset(
#' .dat = in_dat,
#' level1_var, level2_var, level3_var,
#' .count_var = values_var
#' )
#' 
#' draw_sunburst_plot(
#' .dat = plot_dat,
#' .label_txt_size = 4,
#' palette = "Dark2"
#' )
draw_sunburst_plot <- function(
    .dat, 
    .label_txt_size = 1,
    .linewidth = 1,
    ...
){
  
  p <- ggplot2::ggplot(
    data = .dat,
    mapping = ggplot2::aes(
      xmin = .level,
      xmax = .level + 1,
      ymin = .box_min,
      ymax = .box_max,
      fill = .fill,
      label = .name
    )
  ) + 
    ggplot2::geom_rect(
      mapping = ggplot2::aes(
        alpha = .alpha
      ),
      colour = ifelse(is.na(.dat$.fill), NA, "gray90"),
      linewidth = .linewidth
    ) + 
    ggplot2::geom_text(
      mapping = ggplot2::aes(
        x = .level + 0.5,
        y = (.box_max + .box_min)/2,
        angle = .angle
      ),
      size = .label_txt_size
    ) + 
    ggplot2::coord_polar(
      theta = "y"
    ) +
    ggplot2::scale_fill_brewer(
      ..., 
      na.translate = F) + 
    ggplot2::theme_void() +
    ggplot2::theme(
      legend.position = "none"
    )
  
  return(p)
  
}
